### ISSUE TITLE
Make IAMPolicyMember mutable

### ISSUE THREAD

**Issue by yhrn** [GoogleCloudPlatform/k8s-config-connector#196]
IAMPolicyMember is currently immutable and this often breaks declarative Git-based workflows. 

For example, if you have a workflow where you just commit KRM to a repo and it then gets applied automatically and you update `spec.member` of an IAMPolicyMember you also need to know if the resources already exists in the cluster and in that case change it's name, effectively deleting the old and creating a new. This is more imperative than declarative and becomes extra cumbersome and non-obvious if the dev workflow involves generating the KRM.

Another case where this is painful is if you need to update the IAMPolicyMember to correct an error and you then need to either first delete it or rename it instead of just being able to apply the fix.

So in short, it would be great if it was possible to update all parts of an IAMPolicyBinding.

**Comment by caieo** [comment]
Hi @yhrn, 
Thanks for requesting this enhancement. If I'm understanding your request correctly, you'd like to have one IAMPolicyMember resource where updating and changing `spec.member` is allowed instead of having to delete/rename the IAMPolicyMember and recreate it. When updating the IAMPolicyMember, are you expecting the pre-existing IAMPolicyMember to delete that member and replace it with the new one or just keep adding IAMPolicyMembers to the IAMPolicyBinding?

**Comment by yhrn** [comment]
Hi @caieo,

I would expect the pre-existing member to be deleted in that case. It would also be nice if the entire `spec`, or at least the role, was mutable.

**Comment by caieo** [comment]
@yhrn, Thanks for describing your use case and expected behavior for me. This is a scenario that we agree should be improved and will look into designing an intuitive solution soon.

**Comment by annademidova** [comment]
Hey @caieo - seconding this issue. In my case, however, it'd be great to add multiple roles to the same IAMPolicyMember.

**Comment by TravRan** [comment]
@caieo Has there been any progress on this enhancement?

**Comment by xiaobaitusi** [comment]
Hi TravRan, we are planning to add merging support into IAM policy resource which you can specify a part list of members and roles that you are interested in. With [server side apply](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply), KCC would honor the list and merge it with the live state.  Once you claim the owner of those entries in the list with server side apply, deleting/modifying/adding items in the list will result changes into the underlying API. Note that by then IAM policy doesn't have to be a full intent of the permission configuration for the resource. 

Do you think that will achieve your purpose?

**Comment by TravRan** [comment]
@xiaobaitusi I believe it does. It also would address the issue below as well.
https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/199

**Comment by maqiuyujoyce** [comment]
Thanks for your confirmation, @TravRan !

**Comment by xiaobaitusi** [comment]
Hi all,

KCC now supports IAMPartialPolicy represents a non-authoritative intent for the associated Google Cloud resource's IAM policy. The specified bindings will be merged together with any existing bindings and audit configs.

Note that IAMPartialPolicy CRD supports inline modification to add/remove/modify role bindings, please give it a try and let us know if you run into any issue.

Thanks

**Comment by yhrn** [comment]
@xiaobaitusi Thanks, we will try that out and report back here.

**Comment by yhrn** [comment]
@xiaobaitusi We just started playing with `IAMPartialPolicy` and realized that it does not support `memberFrom`. This makes it less useful for us. Is there any plan to add that to `IAMPartialPolicy`?

**Comment by xiaobaitusi** [comment]
Hey @yhrn, yes, it's on our backlog. We are looking into adding `memberFrom` similar to `IAMPolicyMember`.

**Comment by diviner524** [comment]
`IAMPartialPolicy` now supports `memberFrom`.

https://cloud.google.com/config-connector/docs/reference/resource-docs/iam/iampartialpolicy#spec

@yhrn Please give it a try and feel free to reopen the issue if you encounter any user friction.
